### ISSUE TITLE
Boundary conditions on VectorFunctionSpace fail when using fieldsplit

### ISSUE THREAD

**Issue by halawiye** [firedrakeproject/firedrake#1467]
I'm working on a problem that requires me to apply a `DirichletBC` to one component of a `VectorFunctionSpace` that is in itself a component of a `MixedFunctionSpace`. Due to the nature of the problem, I have to use different preconditioners/solvers for the different blocks, which I have achieved in the past using PETSc's PCFIELDSPLIT. However, when I try to combine these two aspects, firedrake does not seem to be able to correctly reconstruct the function spaces after the fieldsplit, causing various instances of `IndexError`, `MapValueError` and `RuntimeError`.

Here is a minimal example I was able to produce that demonstrates one such problem:
```                                                                                               
 from firedrake import *
 
 mesh = UnitSquareMesh(32,32)
 X = SpatialCoordinate(mesh)
 
 V = VectorFunctionSpace(mesh, "CG", 1)
 VV = MixedFunctionSpace([V,V])
 
 f = interpolate(as_vector([1.0, sin(X[0]*X[1]*2*pi)]), V)
 
 uu = Function(VV)
 vv = TestFunction(VV)
 
 L = 0 
 for u, v in zip(split(uu), split(vv)):
     L += inner(grad(u), grad(v))*dx - inner(f,v)*dx
 
 bcs = [ 
     #DirichletBC(VV.sub(0), Constant((0.0, 0.0)), "on_boundary"),
     DirichletBC(VV.sub(0).sub(0), 0.0, "on_boundary"),
     DirichletBC(VV.sub(0).sub(1), 0.0, "on_boundary"),
     #DirichletBC(VV.sub(1), Constant((0.0, 0.0)), "on_boundary")
     DirichletBC(VV.sub(1).sub(0), 0.0, "on_boundary"),
     DirichletBC(VV.sub(1).sub(1), 0.0, "on_boundary"),
     ]   
 
 sp = { 
     "mat_type" : "matfree",
     "snes_type" : "newtonls",
     "snes_monitor" : None,
     "snes_converged_reason" : None,
     "ksp_type" : "fgmres",
     "ksp_monitor_true_residual" : None,
     "ksp_max_it" : 10, 
     "pc_type" : "fieldsplit",
     "pc_fieldsplit_type": "schur",
     "pc_fieldsplit_schur_fact_type": "full",
     "pc_fieldsplit_0_fields" : "0",
     "pc_fieldsplit_1_fields" : "1",
     "fieldsplit_0_ksp_type": "gmres",
     "fieldsplit_0_ksp_monitor_true_residual" : None,
     "fieldsplit_0_ksp_converged_reason" : None,
     "fieldsplit_1_ksp_type": "gmres",
     "fieldsplit_1_ksp_monitor_true_residual" : None,
     "fieldsplit_1_ksp_converged_reason" : None
     }   
 
 solve(L==0, uu, bcs=bcs, solver_parameters = sp) 
 #solve(L==0, uu, bcs=bcs)
```
When I run this with the fieldsplit and the `DirichletBC` applied to the `VV.sub(i).sub(j)`, I get a series of the following errors.
```
SystemError: <built-in method insert of list object at 0x7f23eece0f48> returned a result with an error set
Traceback (most recent call last):
  File "libpetsc4py/libpetsc4py.pyx", line 860, in libpetsc4py.MatMult_Python
  File "/scratch/farrellp/install-scripts/firedrake/firedrake-dev-20190507-mpich/src/firedrake/firedrake/matrix_free/operators.py", line 176, in mult
    bc.zero(self._x)
  File "/scratch/farrellp/install-scripts/firedrake/firedrake-dev-20190507-mpich/src/firedrake/firedrake/bcs.py", line 200, in zero
    r = r.sub(idx)
  File "/scratch/farrellp/install-scripts/firedrake/firedrake-dev-20190507-mpich/src/firedrake/firedrake/function.py", line 319, in sub
    return self._components[i]
IndexError: tuple index out of range

The above exception was the direct cause of the following exception:

SystemError: <built-in method insert of list object at 0x7f23eece0f48> returned a result with an error set
Exception ignored in: 'petsc4py.PETSc.traceback'
```

If I either remove the fieldsplit or apply the `DirichletBC` to `VV.sub(i)` instead, the program runs to completion. Unfortunately, neither of these is an option in my actual problem of interest. Is there a way we can faithfully reconstruct these nested subspace structures after a fieldsplit?

**Comment by wence-** [comment]
Just to check, does this problem go away if you say `"mat_type": "aij"` instead of `"mat_type": "matfree"`? This will help to pin down exactly where the bug is.

**Comment by wence-** [comment]
You may need to use `"matfree"`, but this will help to figure out where to fix things.

**Comment by halawiye** [comment]
In the example posted above, setting `"mat_type" : "aij"` does make the problem go away, but I do indeed seemingly need to use `"matfree"` in my problem as it features an R-space block.

**Comment by wence-** [comment]
OK, thanks, I suspect therefore I know where the bug is, I will try and get to it in the next week or so. It is likely going wrong in the `createSubMatrix` call in operators.py (where the row_bcs and col_bcs are created).

**Comment by wence-** [comment]
@halawiye, can you confirm the fix in #1475 works for your use case?

**Comment by halawiye** [comment]
I can confirm that I no longer see errors of this kind, many thanks to you and @ksagiyam!
